### ISSUE TITLE
Should varbit field appear after its associated length field in header?

### ISSUE THREAD

**Issue by zma-barefoot** [p4lang/p4c#1745]
When compiling the P4_14 program below,

```
header_type hd_t {
    fields {
        options        : *;
        length_        : 8;
    }
    length: length_;
    max_length: 60;
}

header hd_t     hd;

parser start {
    extract(hd);
    return ingress;
}

control ingress { }
```

I get this error. I'm assuming it means `length_` needs to appear before `options` in the header?

```
issue_2.p4(6): [--Werror=not-found] error: length_: Not found declaration
    length: length_;
            ^^^^^^^
```

The equivalent program in P4_16, compiles fine however.

```
#include <v1model.p4>

header hd_t {
    varbit<320> options;
    bit<8>  length;
}

struct packet_t {
    hd_t  hd;
}

struct user_metadata_t {
    bit<8>  data;
}

parser TopParser(packet_in b, out packet_t p,
                 inout user_metadata_t m, inout standard_metadata_t meta) {
    state start {
        b.extract(p.hd, (bit<32>)(p.hd.length));
        transition accept;
    }
}
```

Q: Should varbit field appear after its associated length field in header?

One would assume `length` should always be a priori to `options`, since it comes in first "on the wire"?

**Comment by jafingerhut** [comment]
Out of curiosity, is there some actual packet header format that you are trying to implement from some standard or proposed standard here?

I ask because I do not recall seeing any where the encoding of the length is at the end of the variable-length portion.

I have seen many examples where the encoding of the length is somewhere before the variable-length portion, and you should be able to write a P4_14 or P4_16 program that can implement that just fine.

I doubt there are any P4 implementations that can do what you are trying to do.

**Comment by jafingerhut** [comment]
I made a little P4_16 program out of this that has a debug table so that I could see what is happening when sending a packet into BMv2 simple_switch with that program.  As far as I can tell, it is doing nothing useful.

If there is any issue here, it is that p4c is not giving an error message for the P4_16 program.  It seems very reasonable for p4c to give an error message for any P4_16 program where there is a varbit field in a header that is before any field used in calculating the length of that varbit field.

I suspect that BMv2 simple_switch might only work as intended if the varbit field is last in the header, and if so, p4c should give an error if a varbit field is not last in a header where it occurs.

**Comment by hesingh** [comment]
The issue that p4c p4-16 code needs to catch is "the 2nd arg to extract is an `[in]` arg.  But the value used for this arg is `p.hd.length`.  `p` is an `[out]` parameter.  How can any of members of `p` even have a valid value for an `[in]` parameter - isn't it uninitialized memory.

In regards to placement of varbits, as far as one sticks to TLV format (Type Length Value), data struct design is fine.

**Comment by hesingh** [comment]
I changed p4c code and now p4c catches the two errors discussed thus far for p4-16.
```
hemant@ubuntu:/tmp/p4c/build$ ./p4test --std p4-16 foo1.p4
foo1.p4(19): [--Werror=type-error] error: p.hd: argument should have varbit as last field
        b.extract(p.hd, (bit<32>)(p.hd.length));
                  ^^^^
foo1.p4(19): [--Werror=type-error] error: extract 2nd argument cannot be an OUT parameter: p
        b.extract(p.hd, (bit<32>)(p.hd.length));
                                  ^
hemant@ubuntu:/tmp/p4c/build$ 

```

**Comment by hesingh** [comment]
There are the p4c code changes. 

```
diff --git a/frontends/p4/typeChecking/typeChecker.cpp b/frontends/p4/typeChecking/typeChecker.cpp
index 67aeb26a..f4fc2a0c 100644
--- a/frontends/p4/typeChecking/typeChecker.cpp
+++ b/frontends/p4/typeChecking/typeChecker.cpp
@@ -99,6 +99,20 @@ TypeChecking::TypeChecking(ReferenceMap* refMap, TypeMap* typeMap,
 
 //////////////////////////////////////////////////////////////////////////
 
+/* helper function to get the 'outermost' containing expression in an lvalue */
+static const IR::Expression *lvalue_out(const IR::Expression *exp) {
+    if (auto mem = exp->to<IR::Cast>())
+        return lvalue_out(mem->expr);
+    if (auto mem = exp->to<IR::Member>())
+        return lvalue_out(mem->expr);
+
+    return exp;
+}
+
+static const IR::Path* lvalue_var(const IR::Expression *exp) {
+    return lvalue_out(exp)->to<IR::PathExpression>()->path;
+}
+
 // Make a clone of the type where all type variables in
 // the type parameters are replaced with fresh ones.
 // This should only be applied to canonical types.
@@ -2722,6 +2736,40 @@ bool TypeInference::hasVarbitsOrUnions(const IR::Type* type) const {
     return false;
 }
 
+// Returns true if Varbits field of a header is not the
+// last field.
+bool TypeInference::hasErrVarbits(const IR::Type* type) const {
+    if (type->is<IR::Type_Header>()) {
+        bool hit_varbit = false;
+        auto ht = type->to<IR::Type_Header>();
+        for (auto field : ht->fields) {
+            if (field->type->is<IR::Type_Varbits>()) {
+                hit_varbit = true;
+                continue;
+            }
+            if (hit_varbit)
+                return true;
+        }
+    }
+    return false;
+}
+
+// Returns valid path if 2nd arg of extract is an [out] param which is incorrect.
+const IR::Path* TypeInference::hasOutVarParam(const IR::MethodCallExpression* mce) const {
+  auto arg1 = mce->arguments->at(1);
+  auto exp = arg1->expression;
+  if ((exp->is<IR::Cast>() || exp->is<IR::Expression>()) && !exp->is<IR::Constant>()) {
+      auto lhs = lvalue_var(exp);
+      auto decl = refMap->getDeclaration(lhs, true);
+      if (decl->is<IR::Parameter>()) {
+          auto paramDecl = decl->to<IR::Parameter>();
+          if (paramDecl->direction == IR::Direction::Out)
+              return lhs;
+      }
+  }
+  return nullptr;
+}
+
 bool TypeInference::onlyBitsOrBitStructs(const IR::Type* type) const {
     // called for a canonical type
     if (type->is<IR::Type_Bits>() || type->is<IR::Type_Boolean>() ||
@@ -2798,6 +2846,11 @@ void TypeInference::checkCorelibMethods(const ExternMethod* em) const {
             } else if (argCount == 2) {
                 if (!hasVarbitsOrUnions(argType))
                     typeError("%1%: argument should contain a varbit field", arg0);
+                if (hasErrVarbits(argType))
+                    typeError("%1%: argument should have varbit as last field", arg0);
+                if (auto path = hasOutVarParam(mce))
+                    typeError("extract 2nd argument cannot be an OUT parameter: %1%",
+                              path);
             } else {
                 // core.p4 is corrupted.
                 typeError("%1%: Expected 1 or 2 arguments for '%2%' method",
diff --git a/frontends/p4/typeChecking/typeChecker.h b/frontends/p4/typeChecking/typeChecker.h
index 84a3b6b0..cd8783ed 100644
--- a/frontends/p4/typeChecking/typeChecker.h
+++ b/frontends/p4/typeChecking/typeChecker.h
@@ -129,6 +129,8 @@ class TypeInference : public Transform {
 
     // various helpers
     bool hasVarbitsOrUnions(const IR::Type* type) const;
+    bool hasErrVarbits(const IR::Type* type) const;
+    const IR::Path* hasOutVarParam(const IR::MethodCallExpression* mce) const;
     bool onlyBitsOrBitStructs(const IR::Type* type) const;
     void checkCorelibMethods(const ExternMethod* em) const;
     void checkEmitType(const IR::Expression* emit, const IR::Type* type) const;

```

**Comment by jafingerhut** [comment]
Actually, isn't it legal to read an `out` direction parameter inside of a control or action?  You _should_ initialize it before reading it, but I didn't think it was an error to read it.

**Comment by hesingh** [comment]
Agree.  

However, it is parser code and there is no way p or p.hd is initialized unless the extract operation completes.  This is why p4c catching this error makes sense.

**Comment by jafingerhut** [comment]
An earlier parser state or an earlier assignment statement before the extract call could potentially assign a value to the second argument of extract, even if it is part of `out` struct `p` in this example code.

That would be completely legal.  Would your changes give an error for using _any_ field of `p` as the second arg to extract, even if it had been legally assigned a value in a parser state executed earlier?  If so, then the error message seems incorrect.

**Comment by jafingerhut** [comment]
In fact, the _common_ case for calling 2-variable extract will be that the second argument is either a value that was initialized via an earlier extract() or lookahead() method call, or some expression that includes one or more values initialized in one of those ways.

One thing functionally wrong with the original example code is that the length has not been initialized before the call to the 2-arg version of extract.   The common ways to do so are as mentioned above: an earlier extract or lookahead method call.

**Comment by hesingh** [comment]
Very good.  If the user does not initalize length, I wonder how does extract behave when tested on an actual switch? I expect extract to fail or crash the switch due to uninitialized memory passed to it.  

Does it make sense that if 2nd arg to extract is `p.hd.length`, p4c checks if p.isValid()?  If p is Invalid, p4c raises an error?  But a user could use `p.hd.length = 255` before calling extract, p.hd is Invalid, but length is not.   

Bottom line is if extract can cause a crash at run-time, p4c needs to catch the issue at compile-time

**Comment by hanw** [comment]
It is unlikely the extract will crash the switch, but the packet may be dropped.

**Comment by jafingerhut** [comment]
Detecting bugs statically is a fine thing to do, if it can be done without too much complexity in the compiler.  A warning that the second extract parameter may be used uninitialized in this case may be easy for the compiler to determine -- I am not familiar enough with its internals to say, although it does give such warnings in some cases.

There are also tools like `p4v`  ( https://dl.acm.org/authorize.cfm?key=N666968  a SIGCOMM 2018 paper) that can detect such things statically, and I have not tried it but would guess that p4pktgen could also do so, as it can often detect reads of uninitialized variables ( https://github.com/p4pktgen/p4pktgen ).  Most language compilers do not attempt to catch all possible bugs.

**Comment by hesingh** [comment]
Agree - if a check does not complicate the compiler, we could consider adding the check.  

gcc does support catching uninitialized data, though not reliably for some cases.  Likewise, I found out p4c also does the same.  Run this test case. 

`./p4test --std p4-16 ../testdata/p4_16_samples/uninit.p4`

Clearly, extract is not supported by p4c yet for catching use of uninitialized data.

**Comment by hesingh** [comment]
The test program of this issue has no main module.  I backed out any code changes to p4c, then added code from this program in another P4 program with a main module and now p4c does better.  First see the changes to the p4 program and then the test log.

```
diff --git a/testdata/p4_16_samples/uninit.p4 b/testdata/p4_16_samples/uninit.p4
index 52984e25..ef2048a4 100644
--- a/testdata/p4_16_samples/uninit.p4
+++ b/testdata/p4_16_samples/uninit.p4
@@ -20,6 +20,8 @@ header Header {
     bit<32> data1;
     bit<32> data2;
     bit<32> data3;
+    varbit<320> options;
+    bit<8>  length;
 }
 
 extern void func(in Header h);
@@ -31,6 +33,7 @@ parser p1(packet_in p, out Header h) {
     bool c;
     bool d;
     state start {
+        p.extract(h, (bit<32>)(h.length));
         h.data1 = 0;
         func(h);  // uninitialized
         g(h.data2, g(h.data2, h.data2));  // uninitialized
```

```
hemant@ubuntu:/tmp/p4c/build$ ./p4test --std p4-16 ../testdata/p4_16_samples/uninit.p4
../testdata/p4_16_samples/uninit.p4(36): [--Wwarn=uninitialized_use] warning: h.length may be uninitialized
        p.extract(h, (bit<32>)(h.length));
                               ^^^^^^^^
```

I also tested the bmv2 backend for this issue.  The backend generates a JSON file fine with the varbits included in the header.  Now, I am not sure, if we should change p4c to allow varbits only as the last field in a header.

**Comment by hesingh** [comment]
Sorry to waffle on this one.  Having thought more, we should add a check to p4c frontend to generate an error if varbits is not the last field in a header/struct.  This way, the networking TLV format is honored.  Code changes already exist in this Issue to do so.

**Comment by jafingerhut** [comment]
I know that code changes in PRs can take a bit of time for the appropriate folks to review, if they are busy, but it will probably _never_ happen if the only form you submit them in is as a copy/paste of a diff into an issue comment.  I'd recommend creating a PR that refers to this issue.

**Comment by hesingh** [comment]
Understood.  

However, as you know I made the mistake of creating a fork off of p4c and changing the fork for a PR.  Thus, I cannot create any non-master branches to issue another PR.  This is the reason for adding diffs to an Issue so I don't lose them.  I am working on multiple issues.  Soon as my PR is merged, I will be able to use multiple PR branches and stop adding diffs to Issues.

**Comment by hanw** [comment]
look up the `git remote` command. Do the following to pull the changes in your fork back to a p4c branch.
```
git clone git@github.com:p4lang/p4c.git p4c-repo
cd p4c-repo
git remote add familysingh git@github.com:familysingh/p4c.git
git fetch familysingh
git checkout familysingh/master -b familysingh/master
git push origin hermant/master
```

Here is how what `git remote -v` returns in my local repo.

```
hesingh	git@github.com:hesingh/p4c-1.git (fetch)
hesingh	git@github.com:hesingh/p4c-1.git (push)
mihai	git@github.com:mbudiu-vmw/p4c-clone.git (fetch)
mihai	git@github.com:mbudiu-vmw/p4c-clone.git (push)
origin	git@github.com:p4lang/p4c.git (fetch)
origin	git@github.com:p4lang/p4c.git (push)
singh	git@github.com:familysingh/p4c.git (fetch)
singh	git@github.com:familysingh/p4c.git (push)
```

**Comment by hesingh** [comment]
> look up the `git remote` command. Do the following to pull the changes in your fork back to a p4c branch.
> 
> ```
> git clone git@github.com:p4lang/p4c.git p4c-repo
> cd p4c-repo
> git remote add familysingh git@github.com:familysingh/p4c.git
> git fetch familysingh
> git checkout familysingh/master -b familysingh/master
> git push origin hermant/master
> ```
> Here is how what `git remote -v` returns in my local repo.
> 
> ```
> hesingh	git@github.com:hesingh/p4c-1.git (fetch)
> hesingh	git@github.com:hesingh/p4c-1.git (push)
> mihai	git@github.com:mbudiu-vmw/p4c-clone.git (fetch)
> mihai	git@github.com:mbudiu-vmw/p4c-clone.git (push)
> origin	git@github.com:p4lang/p4c.git (fetch)
> origin	git@github.com:p4lang/p4c.git (push)
> singh	git@github.com:familysingh/p4c.git (fetch)
> singh	git@github.com:familysingh/p4c.git (push)
> ```

Thanks, Han.  I don't understand this command in your list

`> git push origin hermant/master`

I have github userid's in hesingh and familysingh.  There is no hemant.  So what do I replace "hermant" with in the above command?

**Comment by hanw** [comment]
This last command creates a branch in p4lang/p4c repo with a branch name `herman/master`, you can replace the branch name with any string. Typically, we follow the branch naming scheme of `user-name/branch-name`. `origin` is the name of the remote to git@github.com:p4lang/p4c.git

**Comment by hemant-avatar3** [comment]
> I know that code changes in PRs can take a bit of time for the appropriate folks to review, if they are busy, but it will probably _never_ happen if the only form you submit them in is as a copy/paste of a diff into an issue comment. I'd recommend creating a PR that refers to this issue.

I created my 3rd account in github and now using a p4lang/p4c fork with non-master branches.  A PR has been issued:

https://github.com/p4lang/p4c/pull/1748

**Comment by hesingh** [comment]
Should we Close this Issue now that closure has been documented in the PR in the note above?

**Comment by mbudiu-vmw** [comment]
No, the translator p4-14 -> p4-16 should give a better error message.

**Comment by mbudiu-vmw** [comment]
The P4-16 program generated from this one looks like this:
```
header hd_t {
    @length((bit<32>)length_ * 8 - 8) 
    varbit<472> options;
    bit<8>      length_;
}
```

In this header the length_ field is not yet declared when first referenced. That's why you get the error, which is not very clear in the context of the P4-14 program.

The program you wrote in P4-16 is not equivalent. The correct program would look like this:
```
header hd_t_1 {
    bit<8> length_;
}

header hd_t {
    bit<8>      length_;
    @length((bit<32>)length_ * 8 - 8) 
    varbit<472> options;
}

struct headers {
    @name(".hd") 
    hd_t hd;
}

parser ParserImpl(packet_in packet, out headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    hd_t_1 tmp_hdr;
    @name(".start") state start {
        tmp_hdr = packet.lookahead<hd_t_1>();
        packet.extract(hdr.hd, (bit<32>)((bit<32>)tmp_hdr.length_ * 8 - 8));
        transition accept;
    }
}
```

Notice how the program does a lookahead to extract the length field which is next used to extract the full header.

**Comment by mbudiu-vmw** [comment]
There is second potential bug here that the compiler does not warn about using an uninitialized variable. The def-use analysis should have figured out that p.hd.length is not initialized in the call to extract. I will take a look at this.

**Comment by mbudiu-vmw** [comment]
Actually there is no bug; since the program has no main, it is entirely dead code. If one adds a main package the compiler gives a nice error message:
```
issue1745-1.p4(14): [--Wwarn=uninitialized_use] warning: p.hd.length may be uninitialized
        b.extract(p.hd, (bit<32>)(p.hd.length));
                                  ^^^^^^^^^^^
```

**Comment by hesingh** [comment]
Right, this is what I also said yesterday and added the same log to this Issue.
